### ISSUE TITLE
Incorrect generated FormData when the request is a polymorphic type (OneOf schema) 

### ISSUE THREAD

**Issue by ahmed2m** [anymaniax/orval#844]
Recommended to checkout the OpenAPI spec for inheritance and polymorphism before the ticket itself: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

## What are the steps to reproduce this issue?

1. Use the following OpenAPI specification
<details>
  <summary>
    SCHEMA (very long)
  </summary>

```json
{
    "openapi": "3.0.3",
    "info": {
        "title": "Your Project API",
        "version": "1.0.0",
        "description": "Your project description"
    },
    "paths": {
        "/labels/": {
            "get": {
                "operationId": "labels_list",
                "tags": [
                    "labels"
                ],
                "security": [
                    {
                        "cookieAuth": []
                    },
                    {
                        "basicAuth": []
                    }
                ],
                "responses": {
                    "200": {
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "array",
                                    "items": {
                                        "$ref": "#/components/schemas/LabelPolymorphic"
                                    }
                                }
                            }
                        },
                        "description": ""
                    }
                }
            },
            "post": {
                "operationId": "labels_create",
                "tags": [
                    "labels"
                ],
                "requestBody": {
                    "content": {
                        "multipart/form-data": {
                            "schema": {
                                "$ref": "#/components/schemas/LabelPolymorphicRequest"
                            }
                        }
                    }
                },
                "security": [
                    {
                        "cookieAuth": []
                    },
                    {
                        "basicAuth": []
                    }
                ],
                "responses": {
                    "201": {
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/LabelPolymorphic"
                                }
                            }
                        },
                        "description": ""
                    }
                }
            }
        },
        "/tasks/": {
            "get": {
                "operationId": "tasks_list",
                "tags": [
                    "tasks"
                ],
                "security": [
                    {
                        "cookieAuth": []
                    },
                    {
                        "basicAuth": []
                    }
                ],
                "responses": {
                    "200": {
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "array",
                                    "items": {
                                        "$ref": "#/components/schemas/TaskPolymorphic"
                                    }
                                }
                            }
                        },
                        "description": ""
                    }
                }
            },
            "post": {
                "operationId": "tasks_create",
                "tags": [
                    "tasks"
                ],
                "requestBody": {
                    "content": {
                        "multipart/form-data": {
                            "schema": {
                                "$ref": "#/components/schemas/TaskPolymorphicRequest"
                            }
                        }
                    }
                },
                "security": [
                    {
                        "cookieAuth": []
                    },
                    {
                        "basicAuth": []
                    }
                ],
                "responses": {
                    "201": {
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/TaskPolymorphic"
                                }
                            }
                        },
                        "description": ""
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "Label": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphic"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    }
                },
                "required": [
                    "id",
                    "task"
                ]
            },
            "LabelPolymorphic": {
                "oneOf": [
                    {
                        "$ref": "#/components/schemas/LabelTyped"
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionLabelTyped"
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionLabelTyped"
                    }
                ],
                "discriminator": {
                    "propertyName": "resourcetype",
                    "mapping": {
                        "Label": "#/components/schemas/LabelTyped",
                        "TranscriptionLabel": "#/components/schemas/TranscriptionLabelTyped",
                        "TextEditionLabel": "#/components/schemas/TextEditionLabelTyped"
                    }
                }
            },
            "LabelPolymorphicRequest": {
                "oneOf": [
                    {
                        "$ref": "#/components/schemas/LabelTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionLabelTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionLabelTypedRequest"
                    }
                ],
                "discriminator": {
                    "propertyName": "resourcetype",
                    "mapping": {
                        "Label": "#/components/schemas/LabelTypedRequest",
                        "TranscriptionLabel": "#/components/schemas/TranscriptionLabelTypedRequest",
                        "TextEditionLabel": "#/components/schemas/TextEditionLabelTypedRequest"
                    }
                }
            },
            "LabelRequest": {
                "type": "object",
                "properties": {
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphicRequest"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "minLength": 1,
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    }
                },
                "required": [
                    "task"
                ]
            },
            "LabelTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/Label"
                    }
                ]
            },
            "LabelTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/LabelRequest"
                    }
                ]
            },
            "Task": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "uri",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    },
                    "created_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    },
                    "updated_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    }
                },
                "required": [
                    "created_at",
                    "id",
                    "name",
                    "updated_at"
                ]
            },
            "TaskPolymorphic": {
                "oneOf": [
                    {
                        "$ref": "#/components/schemas/TaskTyped"
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionTaskTyped"
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionTaskTyped"
                    }
                ],
                "discriminator": {
                    "propertyName": "resourcetype",
                    "mapping": {
                        "Task": "#/components/schemas/TaskTyped",
                        "TranscriptionTask": "#/components/schemas/TranscriptionTaskTyped",
                        "TextEditionTask": "#/components/schemas/TextEditionTaskTyped"
                    }
                }
            },
            "TaskPolymorphicRequest": {
                "oneOf": [
                    {
                        "$ref": "#/components/schemas/TaskTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionTaskTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionTaskTypedRequest"
                    }
                ],
                "discriminator": {
                    "propertyName": "resourcetype",
                    "mapping": {
                        "Task": "#/components/schemas/TaskTypedRequest",
                        "TranscriptionTask": "#/components/schemas/TranscriptionTaskTypedRequest",
                        "TextEditionTask": "#/components/schemas/TextEditionTaskTypedRequest"
                    }
                }
            },
            "TaskRequest": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "binary",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    }
                },
                "required": [
                    "name"
                ]
            },
            "TaskTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/Task"
                    }
                ]
            },
            "TaskTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TaskRequest"
                    }
                ]
            },
            "TextEditionLabel": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphic"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    },
                    "edited_text": {
                        "type": "string"
                    }
                },
                "required": [
                    "id",
                    "task"
                ]
            },
            "TextEditionLabelRequest": {
                "type": "object",
                "properties": {
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphicRequest"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "minLength": 1,
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    },
                    "edited_text": {
                        "type": "string"
                    }
                },
                "required": [
                    "task"
                ]
            },
            "TextEditionLabelTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionLabel"
                    }
                ]
            },
            "TextEditionLabelTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionLabelRequest"
                    }
                ]
            },
            "TextEditionTask": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "uri",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    },
                    "created_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    },
                    "updated_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    },
                    "text_validator": {
                        "type": "string",
                        "maxLength": 500
                    }
                },
                "required": [
                    "created_at",
                    "id",
                    "name",
                    "updated_at"
                ]
            },
            "TextEditionTaskRequest": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "binary",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    },
                    "text_validator": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": 500
                    }
                },
                "required": [
                    "name"
                ]
            },
            "TextEditionTaskTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionTask"
                    }
                ]
            },
            "TextEditionTaskTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionTaskRequest"
                    }
                ]
            },
            "TranscriptionLabel": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphic"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    },
                    "transcript": {
                        "type": "string"
                    }
                },
                "required": [
                    "id",
                    "task"
                ]
            },
            "TranscriptionLabelRequest": {
                "type": "object",
                "properties": {
                    "task": {
                        "$ref": "#/components/schemas/TaskPolymorphicRequest"
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "color": {
                        "type": "string",
                        "minLength": 1,
                        "title": "The rbg() color string, e.g. rgb(255,231,78)",
                        "maxLength": 50
                    },
                    "transcript": {
                        "type": "string"
                    }
                },
                "required": [
                    "task"
                ]
            },
            "TranscriptionLabelTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionLabel"
                    }
                ]
            },
            "TranscriptionLabelTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionLabelRequest"
                    }
                ]
            },
            "TranscriptionTask": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "readOnly": true
                    },
                    "name": {
                        "type": "string",
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "uri",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    },
                    "created_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    },
                    "updated_at": {
                        "type": "string",
                        "format": "date-time",
                        "readOnly": true,
                        "nullable": true
                    },
                    "validator": {
                        "type": "string",
                        "maxLength": 500
                    }
                },
                "required": [
                    "created_at",
                    "id",
                    "name",
                    "updated_at"
                ]
            },
            "TranscriptionTaskRequest": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": 200
                    },
                    "html_guidelines": {
                        "type": "string"
                    },
                    "can_documents_be_invalidated": {
                        "type": "boolean"
                    },
                    "image": {
                        "type": "string",
                        "format": "binary",
                        "nullable": true
                    },
                    "archived": {
                        "type": "boolean"
                    },
                    "validator": {
                        "type": "string",
                        "minLength": 1,
                        "maxLength": 500
                    }
                },
                "required": [
                    "name"
                ]
            },
            "TranscriptionTaskTyped": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionTask"
                    }
                ]
            },
            "TranscriptionTaskTypedRequest": {
                "allOf": [
                    {
                        "type": "object",
                        "properties": {
                            "resourcetype": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "resourcetype"
                        ]
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionTaskRequest"
                    }
                ]
            }
        },
        "securitySchemes": {
            "basicAuth": {
                "type": "http",
                "scheme": "basic"
            },
            "cookieAuth": {
                "type": "apiKey",
                "in": "cookie",
                "name": "sessionid"
            }
        }
    }
}
```
</details>

3. Use Orval to generate hooks for react-query client
4. Run Orval
5. Check `tasksCreate` function

## What happens?

In the schema, `Task` is a polymorphic type (OneOf) and `Label` is a polymorphic type that includes a reference to `Task` (OneOf as a $ref in object properties).

I'll demonstrate with the following example:

```json
            "post": {
                "operationId": "tasks_create",
                "tags": [
                    "tasks"
                ],
                "requestBody": {
                    "content": {
                        "multipart/form-data": {
                            "schema": {
                                "$ref": "#/components/schemas/TaskPolymorphicRequest"
                            }
                        }
                    }
                },
            }
```
And the $ref here is:

```json
            "TaskPolymorphicRequest": {
                "oneOf": [
                    {
                        "$ref": "#/components/schemas/TaskTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TranscriptionTaskTypedRequest"
                    },
                    {
                        "$ref": "#/components/schemas/TextEditionTaskTypedRequest"
                    }
                ],
                "discriminator": {
                    "propertyName": "resourcetype",
                    "mapping": {
                        "Task": "#/components/schemas/TaskTypedRequest",
                        "TranscriptionTask": "#/components/schemas/TranscriptionTaskTypedRequest",
                        "TextEditionTask": "#/components/schemas/TextEditionTaskTypedRequest"
                    }
                }
            }
```

Orval generates the following:

```ts
export const tasksCreate = (
  taskPolymorphicRequest: TaskPolymorphicRequest,
  options?: SecondParameter<typeof apiCustomInstance>
) => {
  const formData = new FormData();
  formData.append("data", taskPolymorphicRequest);

  return apiCustomInstance<TaskPolymorphic>(
    {
      url: `/tasks/`,
      method: "post",
      headers: { "Content-Type": "multipart/form-data" },
      data: formData,
    },
    options
  );
};
```

Where `TaskPolymorphicRequest` is a polymorphic type (OneOf in the schema)

```ts
export type TaskPolymorphicRequest =
  | TaskTypedRequest
  | TranscriptionTaskTypedRequest
  | TextEditionTaskTypedRequest;
```

## What were you expecting to happen?

Normally if `Task` was not polymorphic the generated code would be something like the following, perhaps the generated code should **utilize the discriminator (both property name and mapping) to determine which type to use and build the form data accordingly**?

```ts
export const tasksCreate = (
  task: TaskRequest,
  options?: SecondParameter<typeof apiCustomInstance>
) => {
  const formData = new FormData();

  const { name, html_guidelines, can_documents_be_invalidated, image, archived } = task;

  formData.append("name", name);

  if (html_guidelines !== undefined) {
    formData.append("html_guidelines", html_guidelines);
  }

  if (can_documents_be_invalidated !== undefined) {
    formData.append("can_documents_be_invalidated", can_documents_be_invalidated.toString());
  }

  if (image !== undefined && image !== null) {
    formData.append("image", image);
  }

  if (archived !== undefined) {
    formData.append("archived", archived.toString());
  }

  return apiCustomInstance<TaskPolymorphic>(
    {
      url: `/tasks/`,
      method: "post",
      headers: { "Content-Type": "multipart/form-data" },
      data: formData,
    },
    options
  );
};
```

## Any other comments?

I included in the example "Label" another polymorphic type that includes a reference of the "Task" polymorphic in it (so a nested OneOf in the object schema properties) basically to challenge how would this work. ([we need to go deeper](https://i.kym-cdn.com/photos/images/newsfeed/000/531/557/a88.jpg)). (UPDATE: It seems nested objects wouldn't work at all and needs #811 to be fixed first)

This works fine with a normal JSON request, because (if I an understanding correctly) Orval would just dump whatever sent data in the client request and typescript would raise type errors when needed.

**Comment by ahmed2m** [comment]
I tried to improve the bug description right now, also I tried to navigate the code myself and what I've gotten so far is that a check for `polymorphic` (oneOf or anyOf) would be added to the `getSchemaFormDataAndUrlEncoded` function because the `$ref` is not matching any type check done there but I am having some trouble understanding what to do from there.

**Comment by melloware** [comment]
Yes this seems complicated. I marked it as a Bug

**Comment by melloware** [comment]
Fixed in 6.22.0
